gflownet/gflownet.py

```diff
@@ -256,112 +269,50 @@ def forward_sample(
+        envs: List,
+        actions: List[Tuple],
+        is_forward: bool = True,
     ):
         """
-        Performs a backward action on one environment.
+        Executes the actions of a list on the environments of a list.
 
         Args
         ----
-        env : GFlowNetEnv or derived
-            An instance of the environment
+        envs : list of GFlowNetEnv or derived
+            A list of instances of the environment
 
-        sampling_method : string
-            - model: uses the current backward policy to obtain the sampling probabilities.
-            - uniform: samples uniformly from the parents of the current state.
+        actions : list
+            A list of actions to be executed on each env of envs.
 
-        model : torch model
-            Model to use as policy if sampling_method="policy"
+        is_forward : bool
+            True if sampling is forward. False if backward.
 
         temperature : float
             Temperature to adjust the logits by logits /= temperature
-
-        done : bool
-            If True, it will sample eos action
         """
-        # TODO: If sampling method is policy
-        # Change backward sampling with a mask for parents
-        # As we can use the backward policy to get model(states) but not all of those actions are likely
-        # Need to compute backward_masks, amsk those actions and then get the categorical distribution.
-        parents, parents_a = env.get_parents(env.state, done)
-        if sampling_method == "policy":
-            action_logits = model(self._tfloat(parents))[
-                torch.arange(len(parents)), parents_a
-            ]
-            action_logits /= temperature
-            if all(torch.isfinite(action_logits).flatten()):
-                action_idx = Categorical(logits=action_logits).sample().item()
-            else:
-                if self.logger.debug:
-                    raise ValueError("Action could not be sampled from model!")
-        elif sampling_method == "uniform":
-            action_idx = self.rng.integers(low=0, high=len(parents_a))
+        assert len(envs) == len(actions)
+        if not isinstance(envs, list):
+            envs = [envs]
+        if is_forward:
+            _, actions, valids = zip(
+                *[env.step(action) for env, action in zip(envs, actions)]
+            )
         else:
-            raise NotImplemented
-        action = parents_a[action_idx]
-        env.set_state((parents)[action_idx], done=False)
-        return env, env.state, action, parents, parents_a
+            valids = []
+            for env, action in zip(envs, actions):
+                parents, parents_a = env.get_parents(action=action)
+                if action in parents_a:
+                    state_next = parents[parents_a.index(action)]
+                    env.set_state(state_next, done=False)
+                    env.n_actions -= 1
```

Comment:
Not sure but wouldn;t we need to update `env.done` as well?

Reply:
Yes, but that is already done by `env.set_state(state_next, done=False)` (`done` will be set to `False` inside `set_state`)

Reply:
Right, my bad, got it, thanks!